### PR TITLE
[SPARK-50546][SQL] Add subquery cast support to collation type coercion

### PR DESCRIPTION
**Pull request by stefankandic**
### What changes were proposed in this pull request?
Adding proper support for implicit casting of subqueries in collation type coercion. Currently I only added support for `Project` and `Aggregate` plans in subquery expression, but we might need to add more later.

### Why are the changes needed?
Since `SubqueryExpression` is a special type of expression it is not possible to just cast it into a different type, and this was causing a runtime error. Instead we need to go inside of its plan and cast the plan outputs.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Added new unit tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by stefankandic** [comment]
@stevomitric @mihailom-db please take a look

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala [hunk 1]
```diff
@@ -180,6 +184,36 @@ object CollationTypeCoercion {
...
+
+    val newPlan = subqueryExpression.plan match {
+      case project: Project =>
+        val newProjectList = project.projectList.map(transformNamedExpressions)
+        project.copy(projectList = newProjectList)
+
+      case agg: Aggregate =>
+        val newAggregateExpressions = agg.aggregateExpressions.map(transformNamedExpressions)
+        agg.copy(aggregateExpressions = newAggregateExpressions)
+
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala:210]: Do we need SubqueryAlias here?

**Review by mihailomilosevic2001 — APPROVED** [review]
LGTM apart one small comment, please could you add an example of changed behaviour to the description

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala (hunk 1 above)
Comment (stefankandic) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala:210]: image
I don't think so, this is how the plan will look like if we have an alias for the subquery

**Comment by stefankandic** [comment]
@MaxGekk can you also take a look?

File: sql/core/src/test/scala/org/apache/spark/sql/collation/CollationTypePrecedenceSuite.scala [hunk 2]
```diff
@@ -175,6 +175,115 @@ class CollationTypePrecedenceSuite extends QueryTest with SharedSparkSession {
...
+  test("in subquery expression") {
+    val tableName = "subquery_tbl"
+    withTable(tableName) {
+      sql(
+        s"""
+           |CREATE TABLE $tableName (
+           | c1 STRING COLLATE UTF8_LCASE,
+           | c2 STRING COLLATE UNICODE
+           |) USING $dataSource
+           |""".stripMargin)
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/collation/CollationTypePrecedenceSuite.scala:187]: indentation should be 2 spaces not 3. Please, fix here and below:
```suggestion
          |CREATE TABLE $tableName (
          | c1 STRING COLLATE UTF8_LCASE,
          | c2 STRING COLLATE UNICODE
          |) USING $dataSource
          |""".stripMargin)
```

**Review by MaxGekk — COMMENTED** [review]
LGTM except of a coding style related comment.

File: sql/core/src/test/scala/org/apache/spark/sql/collation/CollationTypePrecedenceSuite.scala [hunk 3]
```diff
@@ -175,6 +175,115 @@ class CollationTypePrecedenceSuite extends QueryTest with SharedSparkSession {
...
+          |""".stripMargin)
+
+      sql(s"INSERT INTO $tableName VALUES ('a', 'A')")
+
+      assertImplicitMismatch(
+        sql(
+          s"""
+             |SELECT * FROM $tableName
+             |WHERE c1 IN (SELECT c2 FROM $tableName)
+             |""".stripMargin))
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/collation/CollationTypePrecedenceSuite.scala:196]: please, fix indentation in all places

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/collation/CollationTypePrecedenceSuite.scala (hunk 3 above)
Comment (stefankandic) [reply sql/core/src/test/scala/org/apache/spark/sql/collation/CollationTypePrecedenceSuite.scala:196]: Why do we have to do this? This is the way intellij formats multi line strings by default, and looking at the code base it seems like aligning the `|` with the third quote is more often used anyways

Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/collation/CollationTypePrecedenceSuite.scala:196]: > This is the way intellij formats multi line strings by default

Just tune IDE once according to the style guide, and the issue will be fixed forever.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @stefankandic and @mihailom-db for review.

**Pull request closed without merging**